### PR TITLE
Java: Password in Java EE configuration files

### PR DESCRIPTION
**Pull request by luchua-bc**
Storing a plaintext password in a configuration file allows anyone who can read the file to access the password-protected resources. Therefore it is a common attack vector.

This is a common issue with many GitHub repositories. The query detects cleartext passwords stored in XML configuration files.

Please consider to merge the PR. Thanks.

### CODE REVIEW

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  </example>
+
+  <references>
+    <li>
+      CWE:
+      <a href="https://cwe.mitre.org/data/definitions/555.html">CWE-555: J2EE Misconfiguration: Plaintext Password in Configuration File</a>
+    </li>
+    <li>
+      RedHat Security Guide:
+      <a href="https://access.redhat.com/documentation/en-us/jboss_enterprise_application_platform/6.1/html/security_guide/Store_and_Retrieve_Encrypted_Sensitive_Strings_in_the_Java_Keystore">STORE AND RETRIEVE ENCRYPTED SENSITIVE STRINGS IN THE JAVA KEYSTORE</a>
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp:33]: Maybe it would be good to not use all-caps here. It appears the RedHat guide only uses a style which makes the text appear all-caps. The text actually has the following capitalization (can also be seen in the browser tab title):
```suggestion
      <a href="https://access.redhat.com/documentation/en-us/jboss_enterprise_application_platform/6.1/html/security_guide/Store_and_Retrieve_Encrypted_Sensitive_Strings_in_the_Java_Keystore">Store and Retrieve Encrypted Sensitive Strings in the Java Keystore</a>
```

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  |
+    b.getElement() = a.getElement() and
+    a.getName().toLowerCase() = "name" and
+    a.getValue().toLowerCase() in ["password", "pwd"] and
+    b.getName().toLowerCase() = "value" and
+    not isNotPassword(b)
+  )
+  or
+  a.getValue().regexpMatch("(?is).*(pwd|password)\\s*=(?!\\s*;).*") // Attribute value matches password pattern
+select a, "Avoid plaintext passwords in configuration files."
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql:37]: This should probably describe the problem, i.e.:
```suggestion
select a, "Plaintext passwords in configuration files."
```

See also [query metadata style guide](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md#alert-messages):
> The message should factually describe the problem that is being highlighted–it should not contain recommendations about how to fix the problem or value judgements.

File: java/ql/src/experimental/Security/CWE/CWE-555/context.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Context>
+    <!-- BAD: Password of datasource is in not encrypted -->
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-555/context.xml:3]:
```suggestion
    <!-- BAD: Password of datasource is not encrypted -->
```

File: java/ql/src/experimental/Security/CWE/CWE-555/context.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    <Resource name="jdbc/exampleDS" auth="Container" type="javax.sql.DataSource"
+               maxTotal="100" maxIdle="30" maxWaitMillis="10000"
+               username="root" password="1234"
+               driverClassName="com.mysql.jdbc.Driver"
+               url="jdbc:mysql://www.example.com:3306/proj"/>
+
+    <!-- GOOD: Password is encrypted and stored in a password vault -->
+    <Resource name="jdbc/exampleDS" auth="Container" type="javax.sql.DataSource"
+               maxTotal="100" maxIdle="30" maxWaitMillis="10000"
+               username="root" password="${VAULT::exampleDS::password::N2NhZDYzOTMtNWE0OS00ZGQ0LWE4MmEtMWNlMDMyNDdmNmI2TElORV9CUkVBS3ZhdWx0}"
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-555/context.xml:13]: Question to maintainers: Is the "Code scanning" action not working correctly? It shows a lot of false positives:
Query warning screenshot

File: java/ql/test/experimental/query-tests/security/CWE-555/applicationContext.xml [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans
+           http://www.springframework.org/schema/beans/spring-beans-2.5.xsd">
+
+<bean id="dataSource" class="org.apache.commons.dbcp.BasicDataSource" destroy-method="close">
```
Comment (Marcono1234) [inline java/ql/test/experimental/query-tests/security/CWE-555/applicationContext.xml:7]: Could you please format this XML file? It appears that (not only in this line) the indentation is not consistent.

File: java/ql/test/experimental/query-tests/security/CWE-555/context.xml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Context>
+    <!-- BAD: Password of datasource is in not encrypted -->
```
Comment (Marcono1234) [inline java/ql/test/experimental/query-tests/security/CWE-555/context.xml:3]:
```suggestion
    <!-- BAD: Password of datasource is not encrypted -->
```

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp (hunk 1 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp]: Thanks @Marcono1234 for reviewing this PR. I've made the change.

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql (hunk 2 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql]: I've made the change.

File: java/ql/src/experimental/Security/CWE/CWE-555/context.xml (hunk 3 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/context.xml]: I've fixed the typo.

File: java/ql/src/experimental/Security/CWE/CWE-555/context.xml (hunk 4 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/context.xml:13]: I don't know how the code scanning action works either. 

I didn't find the query to detect password in configuration files within the Java codeql repository while I noticed there are explicit queries for .NET and JavaScript in codeql repositories so I thought a query for Java would be helpful. 

However, this query does help to detect false positives like encrypted passwords and password placeholders while these scenarios are common in Java EE deployment. This query can be further fine tuned for Java EE specific requirements.

File: java/ql/test/experimental/query-tests/security/CWE-555/applicationContext.xml (hunk 5 above)
Comment (luchua-bc) [reply java/ql/test/experimental/query-tests/security/CWE-555/applicationContext.xml]: I've formatted the XML file.

File: java/ql/test/experimental/query-tests/security/CWE-555/context.xml (hunk 6 above)
Comment (luchua-bc) [reply java/ql/test/experimental/query-tests/security/CWE-555/context.xml]: I've fixed the typo.

File: java/ql/src/experimental/Security/CWE/CWE-555/context.xml (hunk 4 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/context.xml:13]: After a second thought, I think the code scanning action is using the query in the .NET repository to scan this PR since there is another false positive complaining <code>Generic catch clause</code> with <code>csharp/extractor/Semmle.Extraction.CIL/Entities/Method.cs</code> so it seems the tool falsely processed this Java query as a .NET query.

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  |
+    b.getElement() = a.getElement() and
+    a.getName().toLowerCase() = "name" and
+    a.getValue().toLowerCase() in ["password", "pwd"] and
+    b.getName().toLowerCase() = "value" and
+    not isNotPassword(b)
+  )
+  or
+  a.getValue().regexpMatch("(?is).*(pwd|password)\\s*=(?!\\s*;).*") // Attribute value matches password pattern
+select a, "Plaintext passwords in configuration files."
```
Comment (Marcono1234) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql:37]: Maybe it would also be good to change this to singular "password" and "file":
```suggestion
select a, "Plaintext password in configuration file."
```

File: java/ql/test/experimental/query-tests/security/CWE-555/context.xml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Context>
+    <!-- BAD: Password of datasource is not encrypted -->
+    <Resource name="jdbc/exampleDS1" auth="Container" type="javax.sql.DataSource" maxTotal="100" maxIdle="30" maxWaitMillis="10000" username="root" password="1234" driverClassName="com.mysql.jdbc.Driver" url="jdbc:mysql://www.example1.com:3306/proj" />
```
Comment (Marcono1234) [inline java/ql/test/experimental/query-tests/security/CWE-555/context.xml:4]: Maybe it would be good to revert the changes to this file. The formatting here was fine, and in my opinion it is easier to read when the attributes are distributed over multiple lines.

File: java/ql/src/experimental/Security/CWE/CWE-555/context.xml (hunk 4 above)
Comment (Marcono1234) [reply java/ql/src/experimental/Security/CWE/CWE-555/context.xml:13]: Ah yes that makes sense, it looks like this is indeed the C# query [PasswordInConfigurationFile.ql](https://github.com/github/codeql/blob/main/csharp/ql/src/Configuration/PasswordInConfigurationFile.ql).

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql (hunk 7 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql]: I've made the text change.

File: java/ql/test/experimental/query-tests/security/CWE-555/context.xml (hunk 8 above)
Comment (luchua-bc) [reply java/ql/test/experimental/query-tests/security/CWE-555/context.xml]: Agreed. I've reverted the changes for better readability.

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp [hunk 9]
```diff
@@ -0,0 +1,40 @@
+<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>
+      Storing a plaintext password in a configuration file allows anyone who can read the file to access the password-protected resources. Therefore it is a common attack vector.
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp:5]:
```suggestion
      Storing a plaintext password in a configuration file allows anyone who can read the file to access the password-protected resources.
```

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+import java
+
+predicate isNotPassword(XMLAttribute a) {
+  a.getValue() = "" // Empty string
+  or
+  a.getValue().regexpMatch("\\$\\{.*\\}") // Variable placeholder ${password}
+  or
+  a.getValue().charAt(a.getValue().length() - 1) = "=" // A basic check of encrypted passwords ending with padding characters, which could be improved to be more accurate.
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql:19]:
```suggestion
  a.getValue().matches("%=") // A basic check of encrypted passwords ending with padding characters, which could be improved to be more accurate.
```

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+from XMLAttribute a
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql:22]: Give `a` and `b` better names (e.g. `nameAttr`, `valueAttr`)

**Comment by smowton** [comment]
I note from the code scanning warnings (which come from the C# analysis) that this is a near-copy of https://github.com/github/codeql/blob/main/csharp/ql/src/Configuration/PasswordInConfigurationFile.ql -- @github/codeql-csharp, what do you make of the differences (stronger `isNotPassword` predicate vs. the C# version's third disjunct `a.getValue().regexpMatch("(?is).*(pwd|password)\\s*=(?!\\s*;).*")` whose meaning is hazy to me but I think it accepts strings like "password=dfsdfdsf", plus whatever `(?!\\s*;)` is doing)

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp (hunk 9 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.qhelp]: Thanks @smowton for reviewing this PR. I've made the text change.

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql (hunk 10 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql]: I've updated the query.

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql (hunk 11 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql]: I've updated the variable name.

**Comment by hvitved** [comment]
> @github/codeql-csharp, what do you make of the differences (stronger `isNotPassword` predicate vs. the C# version's third disjunct `a.getValue().regexpMatch("(?is).*(pwd|password)\\s*=(?!\\s*;).*")` whose meaning is hazy to me but I think it accepts strings like "password=dfsdfdsf", plus whatever `(?!\\s*;)` is doing)

Looks like this query also has that disjunct: https://github.com/github/codeql/pull/4814/files#diff-a072f910060223db1a33816cd6374601e46f0bb2759b46e60edd0e83e3c90636R36. Not sure why `isNotPassword` is not applied to all disjuncts.

**Comment by smowton** [comment]
Doh yeah. @luchua-bc can we usefully use `isNotPassword` for the cases not currently covered?

**Comment by luchua-bc** [comment]
Thanks @hvitved for reviewing this PR.

I copied the third disjunct from the C# query. Typically .NET applications have a database configuration like the following:
```
  <connectionStrings>
    <add name="sqlServerConnection" connectionString="Data Source=servername;Database=dbname;User ID=username;Password=dfsdfdsf;Connection Timeout=30;" providerName="System.Data.SqlClient" />
  </connectionStrings>
```
This is the reason for this disjunct.

While Java EE applications have a more structured configuration like:
```
<Resource name="jdbc/exampleDS" auth="Container" type="javax.sql.DataSource"
               maxTotal="100" maxIdle="30" maxWaitMillis="10000"
               username="root" password="1234"
               driverClassName="com.mysql.jdbc.Driver"
               url="jdbc:mysql://www.example.com:3306/proj" />
```

I searched Java GitHub repositories and cannot find any repository with the .NET style configuration of connection strings.

Will it be a good idea to simply delete this disjunct since it won't be useful to most Java projects? I can apply the same <code> isNotPassword</code> predicate to the pattern <code>connection="...Password=dfsdfdsf..."</code> if it is still desired.

Please advise.

Thanks,
@luchua-bc

**Comment by luchua-bc** [comment]
I've modified the query to apply the same check (blank, placeholder and encrypted) to embedded passwords. As method parameters cannot be <code>string</code>, I developed a new method with the parameter <code>XMLAttribute</code>.

Please let me know if this is not what we want.

**Comment by smowton** [comment]
They can be `string` -- do you say that because you got a message relating to binding? If so you might need to write something like the following example, which uses `bindingset` to indicate that the set of matches is bounded if the parameter `name` is itself bounded. This is necessary only when the body of the predicate doesn't make it obviously finite -- for example, `name = ["hello", "world"]` does not need this annotation.

```
/**
   * Holds if the name of this algorithm matches `name` modulo case,
   * white space, dashes and underscores.
   */
  bindingset[name]
  predicate matchesName(string name) {
    exists(name.regexpReplaceAll("[-_]", "").regexpFind("(?i)\\Q" + getName() + "\\E", _, _))
  }
```

See also https://codeql.github.com/docs/ql-language-reference/predicates/#binding-sets

**Comment by luchua-bc** [comment]
Great. Thanks a lot @smowton. I've revamped the functions to take a string parameter, which helps to remove the duplicate code and make the query more clear.

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ * @id java/password-in-configuration
+ * @tags security
+ *       external/cwe/cwe-555
+ *       external/cwe/cwe-256
+ *       external/cwe/cwe-260
+ */
+
+import java
+
+/* Holds if the attribute value is not a cleartext password */
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql:14]: Regular comments should be `/**` doc comments (here and below)

**Review by smowton — COMMENTED** [review]
Looks good, will get a code-owner review

File: java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql (hunk 12 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-555/PasswordInConfigurationFile.ql]: Thanks @smowton. I've made the change.

**Review by aschackmull — APPROVED** [review]
